### ISSUE TITLE
Update of PPTV get key algorithm

### ISSUE THREAD

**Issue by lamlam123tld3** [soimort/you-get#1957]
The code for getting the key in pptv.py is not so human-friendly, which actually is a TEA Encryption of the current unix timestamp (UTC +8).

Here is a PHP way of getKey:

/* This key may be important as it can be used to acquire a lot of interesting things.*/

```
date_default_timezone_set("Asia/Shanghai");

function U8($i) {
	return $i & 0xFF;
}

function U32($i) {
	return $i & 0xFFFFFFFF;
}

function pptv_tea_encrypt($data, $k) {
				
	$byte2int = function($bytes) {
		return U32((U8($bytes[3])<<24) + (U8($bytes[2])<<16) + (U8($bytes[1])<<8) + U8($bytes[0]));
	};
	$int2byte = function($int){
		return unpack("C4", pack("L", U32($int)));
	};
	
	$rounds = 32; // 32 rounds encryption
	$delta = 0x9E3779B9; // Golden ratio
	
	[$k0, $k1, $k2, $k3] = $k;
	
	$sum = 0;
	$v0 = $byte2int(array_slice($data, 0, 4));
	$v1 = $byte2int(array_slice($data, 4, 4));
	
	for ($i = 0; $i < $rounds; $i++) {
		$sum = U32($sum + $delta);
		$v0 = U32($v0 + (U32(U32(U32($v1 << 4) + $k0) ^ U32($v1 + $sum)) ^ U32(U32($v1 >> 5) + $k1)));
		$v1 = U32($v1 + (U32(U32(U32($v0 << 4) + $k2) ^ U32($v0 + $sum)) ^ U32(U32($v0 >> 5) + $k3)));
	}
	
	array_splice($data, 0, 4, $int2byte($v0));
	array_splice($data, 4, 4, $int2byte($v1));

        $enc = "";
        for($i = 0; $i < 8; $i++) $enc .= unpack("h2", pack("C", U8($data[$i])))[1];

        return $enc;
}

function pptv_getKey() {
	//TEA Encryption
	
	$data = time();
	$data = array_map("ord", str_split(substr(dechex($data), -8)));
	
	$k = [1896220160, 101056625, 100692230, 7407110]; // static for "qqqqqww"

	return pptv_tea_encrypt($data, $k);
}
```

some usages of key:

some return from "http://web-play.pptv.com/webplay3-0-{$id}.xml?version=4" may be blocked, just add key in the query;
to get mp4 segments, just add key in the query;
to get single mp4, just enter the w entrance with the key in query

**Comment by denobisipsis** [comment]
For the example

Vodplayst="Fri May 26 22:32:19 2017 UTC"

 strtotime($string)-60 is 1495837879.

Then, your code returns                  261558b95a4af4ee
but the value swf returns on site is 43feaf57e8875f76.

I thinks is not standard tea encryption.

This code returns the same value as the swf player

class ckey
{
function shiftright($decimal , $right)
	{
	$decimal&=0xFFFFFFFF;
	$shift=$decimal >> $right;    
	if($decimal >= 0) return $shift;
	return bindec(substr(decbin($shift),$right));
	}
    
function charsToLongs($arg)
	{
	$l2 = array();$l3 = -1;
	
	while (++$l3 < ceil((sizeof($arg) / 4))) 
		{
		$temp=0;
		
		for ($t=0;$t<4;$t++) 
			$temp+=$arg[$l3*4+$t] << $t*8;
		
		$l2[] = $temp;
		}
	return $l2;
	}
	    	
function longsToChars($arg)
	{     
	$l3 = -1;$l2="";
	
	while (++$l3 < sizeof($arg)) 	
		for ($t=0;$t<4;$t++) $l2.=strrev(bin2hex(chr(($arg[$l3] >> 8*$t) & 0xFF)));		
	
	return $l2;
	}  
	    
function getkey($key)
	{
	$l2=array_values(unpack("C*",$key));
	
	$l4=$l3=0;
	
	while($l4<sizeof($l2))	     
		$l3^=$l2[$l4]<<($l4++ %4)*8;
	
	return $l3;	 
	}

function encrypt($string) 
	{
	$key = ckey::getkey("qqqqqww");//1896220160

	for ($k=1;$k<4;$k++)
		${"k$k"} = $key << 8*$k | ckey::shiftright($key , 32 - 8*$k);
		
	$p=ckey::charsToLongs(array_map("ord",str_split(dechex(strtotime($string)-60))));

	$x = 0; $y = -1;
	
	while(++$y < 32)
		{
		$x+= 2654435769;
		
		$t1 = ($p[1] << 4) + $key;
		$t2 = $p[1] + $x;
		$t3 = ckey::shiftright($p[1] , 5) + $k1;
		$p[0]+= $t1 ^ $t2 ^ $t3;		
		
		$t4 = ($p[0] << 4) + $k2;
		$t5 = $p[0] + $x;
		$t6 = ckey::shiftright($p[0] , 5);
		$p[1]+= $t4 ^ $t5 ^ ($t6 + $k3);			
		}

         for ($i=0;$i<8;$i++)		     		     
		$z.=dechex(intval(rand(0,15))).dechex(intval(rand(0,15)));
	    		    				
	return ckey::longsToChars(array($p[0],$p[1])).$z;
	}
}

**Comment by denobisipsis** [comment]
Ok, your code fails in the right shift

$v0 = U32($v0 + (U32(U32(U32($v1 << 4) + $k0) ^ U32($v1 + $sum)) ^ U32(U32($v1 >> 5) + $k1)));
$v1 = U32($v1 + (U32(U32(U32($v0 << 4) + $k2) ^ U32($v0 + $sum)) ^ U32(U32($v0 >> 5) + $k3)));


$v1>>5 and $v0>>5 should be shiftrighted (>>>)

**Comment by airdge** [comment]
这算法应该是错的  运算过程都超出php所能转换的最大值,这边直接做位运算肯定不行
```
ts:1495930471
nodejs:3ff4ea2b42c5f0d50c31b56834b2246a (3ff4ea2b42c5f0d5)
python:3ff4ea2b42c5f0d59d082f9e5a7aa812 (3ff4ea2b42c5f0d5)
php:3ff4ea2b42c5f0d5ed82cb4bf813d827 (3ff4ea2b42c5f0d5)
denobisipsis:3ff4ea2b42c5f0d599792278b3ee013d (3ff4ea2b42c5f0d5)
lamlam123tld3:14c7e9605df16049 
[Finished in 1.1s]
```

同时间戳不同平台计算出来的都是:3ff4ea2b42c5f0d5
```
http://61.179.109.166/w/51e781792fc3653069acfe31622845d4.mp4?type=fpp&key=3ff4ea2b42c5f0d5(正常)
http://61.179.109.166/w/51e781792fc3653069acfe31622845d4.mp4?type=fpp&key=14c7e9605df16049(错误)
```

**Comment by lamlam123tld3** [comment]
the output from my php generate the correct result @denobisipsis @airdge 

>./php.exe ./pptv.php http://v.pptv.com/show/FYtq6XnCKWfKSLA.html
string(16) "3ff4ea2b42c5f0d5"
Title: 中超-17赛季-《中超吐口秀》杨昊预告片：为何他变身拍蚊小能手-专题
Quality: Orignal
Filename: 2e1dce448badbe43ccffd3fbe28c6397.mp4
Size: 35.34M
http://61.244.110.134/w/2e1dce448badbe43ccffd3fbe28c6397.mp4?key=3ff4ea2b42c5f0d5


may be your PHP version is too old?
I am using

>./php.exe  -v
PHP 7.1.2 (cli) (built: Feb 14 2017 21:24:49) ( ZTS MSVC14 (Visual C++ 2015) x64 )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies

**Comment by lamlam123tld3** [comment]
@denobisipsis 
btw, i don't think PHP has this >>> operator

**Comment by airdge** [comment]
@lamlam123tld3 
PHP7运算没问题,PHP5.6出错
PHP7.1
```
ts:1496028692
nodejs:8e2fe25edbb965ded0e1d2108e721bec
python:8e2fe25edbb965de47fa700d79436022
php:8e2fe25edbb965dec9635e93a599907c
denobisipsis:00008a0c008d2316bc940abf29be9214  (403)
lamlam123tld3:8e2fe25edbb965de
```
PHP5.6
```
ts:1496028692
nodejs:8e2fe25edbb965ded66cc9abc4d28bb5
python:8e2fe25edbb965dedc48ecee33cfad60
php:8e2fe25edbb965de8d909f44d65881b8
denobisipsis:8e2fe25edbb965dee4f6c97aea9b0cd8
lamlam123tld3:b9ec8dae342dd7d4 (403)
```

**Comment by lamlam123tld3** [comment]
@airdge 
[$k0, $k1, $k2, $k3] = $k;
Yes,
this type of assigning [a, b] = [b, a] is supported from PHP 7.1

**Comment by denobisipsis** [comment]
Right, its the php version.  Anyway, the case is that the shift right should be unsigned.

**Comment by lamlam123tld3** [comment]
@denobisipsis 
signed or unsigned in this case may only be aware if it is 32-bit php
using 64-bit should not have any problem

**Comment by soimort** [comment]
Closed via #2025.
